### DISCUSSION TITLE
macOS, CMSampleBuffer on the Listen endpoint

### DISCUSSION THREAD

**Discussion by BenLumenDigital** [deepgram/community#150] [General help] [answered]
Hi everyone, looking for help or advice on macOS and how to use a CMSampleBuffer for transcriptions.

I have the iOS demo project working, and I also got that running under native macOS.

But I need to use the CMSampleBuffer based callback, and in all my attempts the Listen endpoint is returning an empty transcript:

`{"channel_index":[0,1],"duration":4.029999,"start":39.65,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"d41df247-2918-4669-9a95-38fb180b46f3","model_info":{"name":"general-nova","version":"2023-03-13.31000","arch":"nova"},"model_uuid":"3312b52b-d08e-4072-bfff-124c410b770d"}}`

I've tried converting the CMSampleBuffer to Data, and tried converting CMSampleBuffer to AVAudioBuffer and then to Data - but both end up with the same result.

I tried appending my converted AVAudioBuffer to a file - and that looks and playsback perfectly. My resulting file is a 48,000hz 32-bit float WAV file.

Any bright ideas out there?!  Thanks.

**Comment by nikolawhallon** [comment, marked as answer]
What are your values for `encoding`, `sample_rate`, and `channels`? I don't think Deepgram supports 32-bit float PCM ***for raw audio streams, that would be my guess as to what's going on here. One can convert 32-bit floats to 16-bit ints (linear16) via something like:

```
    short f32_to_i16(float sample)
    {
        sample = sample * 32768;
        if (sample > 32767)
        {
            return 32767;
        }
        if (sample < -32768)
        {
            return -32768;
        }
        return (short) sample;
    }
```

(sorry this is a random C# example I have on hand, but the logic is analogous). This seems a bit of a hacky fix for now maybe, but at least it does save quite a bit on bandwidth.

**Reply by BenLumenDigital** [reply]
Thank you!  You are correct.  32-bit float was the problem.  a 16-bit Int format does work.  Here's a little helper function incase anyone else comes aross this problem:
```

func pcm32To16(buffer: AVAudioPCMBuffer) -> AVAudioPCMBuffer? {
        let targetFormat = AVAudioFormat(commonFormat: AVAudioCommonFormat.pcmFormatInt16, sampleRate: 48000, channels: 1, interleaved: false)

        if let convertBuffer = AVAudioPCMBuffer(pcmFormat: targetFormat!, frameCapacity: buffer.frameCapacity) {
            convertBuffer.frameLength = convertBuffer.frameCapacity
            if let converter = AVAudioConverter(from: buffer.format, to: convertBuffer.format) {
                converter.sampleRateConverterAlgorithm = AVSampleRateConverterAlgorithm_Normal
                converter.sampleRateConverterQuality = .max
            
                do {
                    try converter.convert(to: convertBuffer, from: buffer)
                } catch let error {
                    print("Conversion error: \(error)")
                    return nil
                }
            }
            
            return convertBuffer
        }
        
        return nil
    }
```
